36113: Support CSRF

labkey-nicka opened the pull request:
- allows support for targeting v19.1+ LabKey instances
- request CSRF from login-whoAmI.api action
- enable cookies on UserAgent to maintain session
- coalesce handling of common parameters, authentication creation,
building urls, and establishing server context

labkey-nicka committed:
CSRF support
- allows support for targeting v19.1+ LabKey instances
- request CSRF from login-whoAmI.api action
- enable cookies on UserAgent to maintain session
- coalesce handling of common parameters, authentication creation,
building urls, and establishing server context

labkey-nicka committed:
CSRF: handle response errors

labkey-nicka committed:
CSRF: Reuse userAgent if user supplied
- can be supplied via -userAgent parameter

labkey-kevink commented on src/lib/LabKey/Query.pm:
```diff
@@ -140,78 +141,37 @@ NOTE:
+        'query.queryName' => $args{'-queryName'}
+    };
+
+	if ($args{'-requiredVersion'}) {
```
Remove the note in docs above selectRows that requiredVersion now defaults to 8.3 to match other LabKey api libraries.

labkey-nicka replied:
Good call, meant to change this back to 9.1 default. Can later be adjusted to 8.3 if it is decided that a major release version is warranted.

labkey-kevink commented on src/lib/LabKey/Query.pm:
```diff
@@ -140,78 +141,37 @@ NOTE:
-			$params{"query.".$_} = $args{'-'.$_};
+	foreach ('viewName', 'offset', 'sort', 'maxRows', 'columns', 'containerFilterName') {
+		if ($args{'-'.$_}) {
+            $$data{'query.'.$_} = $args{'-'.$_};
```
Include spaces around the '.' operator.

```suggestion
            $$data{'query.' . $_} = $args{'-' . $_};
```

labkey-kevink commented on src/lib/LabKey/Query.pm:
```diff
@@ -478,58 +359,27 @@ NOTE:
-	foreach ('offset', 'sort', 'maxRows', 'containerFilterName'){
-		if ( $args{'-'.$_} ) {
+	foreach ('offset', 'sort', 'maxRows', 'containerFilterName') {
+		if ($args{'-'.$_}) {
```
```suggestion
		if ($args{'-' . $_}) {
```

labkey-kevink reviewed (approved):
Looks good

labkey-nicka committed:
selectRows: maintain 9.1 default apiVersion
- will require a major version increment later if this change is made
- linting
